File: src/AppETLCrimeData.scala
package demo

import collection.JavaConverters._
import java.util.UUID.randomUUID
import org.apache.kudu.client._
import org.apache.kudu.spark.kudu.{KuduContext, KuduWriteOptions}
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions.udf
import org.apache.spark.sql.types.{IntegerType, StringType, StructField, StructType}
import org.slf4j.LoggerFactory

object AppETLCrimeData {
  val logger = LoggerFactory.getLogger(AppETLCrimeData.getClass)
  val kuduMasters: String = System.getProperty("KUDU_MASTERS", "cdp:7051")
  val tableName: String = System.getProperty("TABLE_NAME", "default.bpd_crime_data")

  def main(args: Array[String]): Unit = {
    val spark = SparkSession
      .builder()
      .appName("AppETLCrimeData")
      .config("spark.master", "local")
      .config("spark.driver.memory", "4g") 
      .config("spark.executor.memory", "4g")
      .getOrCreate()

    spark.conf.set("spark.sql.files.maxRecordsPerFile", 300000)

    import spark.implicits._

    val kuduContext = new KuduContext(kuduMasters, spark.sqlContext.sparkContext)

    try {
      if (!kuduContext.tableExists(tableName)) {
        throw new RuntimeException(tableName + ": does not exist")
      }
      logger.info(s"Found table: " + tableName)

     val df = spark.read.option("header", "true").option("delimiter", "\t")
                .csv("file:///home/centos/data/BPD_Part_1_Victim_Based_Crime_Data_tab.txt")

     val uuid = udf(() => java.util.UUID.randomUUID().toString)
     val keyedDf = df.withColumn("id", uuid())

     val latitude = (v: String) => {
       if (v == null || v.trim().length()==0)
         -1
       else 
         v.replaceAll("[()]+", "").split(",")(0).trim().toDouble
     }

     val longitude = (v: String) => {
       if (v == null || v.trim().length()==0)
         -1
       else
         v.replaceAll("[()]+", "").split(",")(1).trim().toDouble
     } 

     val year = (v: String) => {
       if (v == null || v.trim().length()==0)
         -1
       else
         v.split("/")(2).trim().toInt
     }

     spark.udf.register("latitude", latitude)
     spark.udf.register("longitude", longitude)
     spark.udf.register("year", year) 

     keyedDf.createOrReplaceTempView("UNPARSED_LAT_LON")

/*
CREATE TABLE bpd_crime_data
(
  id STRING NOT NULL,
  crimeyear INT NOT NULL,
  crimecode STRING,
  crimedate STRING,
  crimetime STRING,
  lat DOUBLE,
  lon DOUBLE,
  address STRING,
  description STRING,
  insideflag STRING,
  weapon STRING,
  post STRING,
  district STRING,
  neighborhood STRING,
  total STRING,
  PRIMARY KEY (id, crimeyear)
)
PARTITION BY HASH(id) PARTITIONS 10,
RANGE (crimeyear)
(
PARTITION 2010 < VALUES <= 2011,
PARTITION 2011 < VALUES <= 2012,
PARTITION 2012 < VALUES <= 2013,
PARTITION 2013 < VALUES <= 2014,
PARTITION 2014 < VALUES <= 2015,
PARTITION 2015 < VALUES <= 2016
)
STORED AS KUDU
TBLPROPERTIES (
'kudu.num_tablet_replicas' = '1'
)
;
*/

     val sql = """
       SELECT id, year(crimedate) AS crimeyear, crimecode, crimedate, crimetime, 
              latitude(latlon) AS lat, longitude(latlon) AS lon,
              location AS address, description, insideflag, weapon, post, district, neighborhood, total
       FROM UNPARSED_LAT_LON
     """
     val finalDf = spark.sql(sql)
     finalDf.printSchema()
     finalDf.show(10)

     val path = "file:///home/centos/data/BPD_Part_1_Victim_Based_Crime_Data_transformed.parquet"
     finalDf.repartition(1).write.partitionBy("crimeyear").parquet(path)

     kuduContext.insertRows(finalDf, tableName, new KuduWriteOptions(ignoreDuplicateRowErrors = true))
     
   }
   catch {
     case unknown : Throwable => logger.error(s"Exception occurred: " + unknown)
   } 
   finally { 
     logger.info(s"Session closing")
     spark.close()
   }
  }
}

File: src/python/AppETLCrimeData.py
import re
import uuid
from pyspark import SparkConf
from pyspark.context import SparkContext
from pyspark.sql.functions import udf
from pyspark.sql.session import SparkSession
from pyspark.sql.types import StringType, IntegerType, DoubleType

def createContext(cores: int, gb: int) -> SparkContext:
  conf = (
    SparkConf()
     .setMaster('local[{}]'.format(cores))
     .setAppName("Cloudera PSE Demo AppETLCrimeData")
     .set('spark.driver.memory', '{}g'.format(gb))
  )
  sc = SparkContext(conf=conf)
  sc.setLogLevel("ERROR")
  return sc

def uuid(v):
  return uuid.uuid4()

def latitude(v: str):
  if (v is None or len(v.strip())==0):
    return -1
  return float(re.sub("[()]+", "", v).split(",")[0].strip())

def longitude(v: str):
  if (v is None or len(v.strip())==0):
    return -1
  return float(re.sub("[()]+", "", v).split(",")[1].strip())

def year(v: str):
  if (v is None or len(v.strip())==0):
    return -1
  return int(v.split("/")[2].strip())

def main():
  spark = SparkSession(createContext(1, 2))
  df = spark.read.option("header", "true").option("delimiter", "\t").csv("s3a://statisticalfx/Data/Public Safety/BPD_Part_1_Victim_Based_Crime_Data_tab.txt")
  df.createOrReplaceTempView("RAW_DATA")

  uuidUDF = udf(lambda v: uuid(v))
  spark.udf.register("uuidUDF", uuid, StringType())

  latUDF = udf(lambda v: latitude(v))
  spark.udf.register("latUDF", latitude, DoubleType())

  lonUDF = udf(lambda v: longitude(v))
  spark.udf.register("lonUDF", longitude, DoubleType())
  
    yearUDF = udf(lambda v: year(v))
  spark.udf.register("yearUDF", year, IntegerType())

  sql = """
       SELECT uuid() AS id, yearUDF(crimedate) AS crimeyear, crimecode, crimedate, crimetime,
              latUDF(latlon) AS lat, lonUDF(latlon) AS lon,
              location AS address, description, insideflag, weapon, post, district, neighborhood, total
       FROM RAW_DATA
  """

  finalDf = spark.sql(sql)
  finalDf.printSchema()
  finalDf.show(10)

if __name__ == "__main__":
  main()
  
  
